brainiak/reprsimil/brsa.py

```diff
@@ -30,24 +32,158 @@
+
+        See also: `.prior_GP_var_half_cauchy`
+
+    Parameters
+    ----------
+    y_invK_y: float
+        y * inv(K) * y^T, where y=f(x) is a vector of observations
+        of unknown function f at different locations x.
+        K is correlation matrix of f between different locations, based
+        on a Gaussian Process (GP) describing the smoothness property
+        of f. K fully incorporates the form of the kernel
+        and the length scale of the GP, but not the variance of the GP
+        (the purpose of this function is to estimate the variance).
+    n_y: int, number of observations
+    tau_range: float,
+        The reasonable range of tau, the standard deviation of the
+        Gaussian Process imposed on y=f(x). tau_range is parameter
+        of the inverse-Gamma prior.
+        The smaller it is, the more penalization is imposed
+        on large variation of y.
+    Returns:
+    --------
+    tau2: The MAP estimation of tau^2 based on the prior on tau
+        and y_invK_y.
+    log_ptau: log(p(tau)) of the returned tau^2 based on the
+        inverse-Gamma prior.
+    """
+    tau2 = (y_invK_y + 2 * tau_range**2) / (2 * 2 + 2 + n_y)
+    log_ptau = scipy.stats.invgamma.logpdf(
+        tau2, scale=tau_range**2, a=2)
+    return tau2, log_ptau
+
+
+def prior_GP_var_half_cauchy(y_invK_y, n_y, tau_range):
+    """ Imposing a half-Cauchy prior onto the standard deviation (tau)
+        of the Gaussian Process which is in turn a prior imposed over
+        a function y = f(x).
+        The scale parameter of the half-Cauchy prior is tau_range.
+        The function returns the MAP estimate of tau^2 and
+        log(p(tau|tau_range)) for the estimated value of tau^2,
+        where tau_range describes the reasonable range of tau
+        in the half-Cauchy prior.
+        An alternative form of prior is inverse-Gamma prior on tau^2.
+        Inverse-Gamma prior penalizes for both very small and very
+        large values of tau, while half-Cauchy prior only penalizes
+        for very large values of tau.
+        For more information on usage, see description in BRSA class:
+        `.BRSA`
+    """
+    tau2 = (y_invK_y - n_y * tau_range**2
+            + np.sqrt(n_y**2 * tau_range**4 + (2 * n_y + 8)
+                      * tau_range**2 * y_invK_y + y_invK_y**2))\
+        / 2 / (n_y + 2)
+    log_ptau = scipy.stats.halfcauchy.logpdf(
+        tau2**0.5, scale=tau_range)
+    return tau2, log_ptau
+
+
+def Ncomp_BIC_Minka(X):
+    """ This implements the BIC metric written in Minka 2000:
+        Automatic choice of dimensionality for PCA
+        The paper mainly proposed a Laplace approximation of the posterior
+        of the number of components but also provided this BIC approximation.
+        For cases in which feature number is smaller than sample size,
+        there is ambiguity for the author of this code to decide how to
+        set some variables. Empirically the number of components estimated
```

Comment:
Sorry about the confusion. If n is the number of observations, and k the number of dimensions, the paper requires k to be smaller than n. 
So the full story is: vanilla implementation of the method often decides to choose k equal to the maximal possible value (rank of the data matrix - 1) for real fMRI data. The approximate log(p(D|M)) given by BIC often appears as a shape of "2" rotated by 90 degrees, with the right-hand tail higher than the peak on the left. This is undesirable and strange. My interpretation is that the BIC approximation works when the number of observation is sufficiently large (as in https://en.wikipedia.org/wiki/Bayesian_information_criterion, although I do not know the correct criterion of calling n "large"). For fMRI data which often have more channels (voxels) than observations, the penalization term might not be sufficient. With a lot of trial-and-error, using larger penalization (n_T * n_V) as the number of observations manages to suppress the tail. But the chosen k is often much smaller than Ncomp_SVHT_MG_DLD_approx gives. Simulation result also shows that the k chosen by Ncomp_SVHT_MG_DLD_approx gives better performance in terms of the recovered similarity matrix.

So, I think I will just remove this BIC metric. I might give a try of the formula not using BIC approximation in the paper. If it fails, Ncomp_SVHT_MG_DLD_approx will be the only option.